hphp/hack/src/hh_codesynthesis/agentGraphGenerator.py

```diff
@@ -0,0 +1,266 @@
+#!/usr/bin/env python3
+# pyre-strict
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the "hack" directory of this source tree.
+#
+# This library will construct a logic program using the given parameter,
+# invoke Clingo to solve the logic program to produce an agent graph,
+# and we create multiple agents in parallel to produce valid codebase.
+import argparse
+import sys
+
+from typing import List, Optional, Set, Union
+
+import clingo
+from clingo.symbol import Number, Symbol
+
+
+class AgentGraphClingoContext:
+    """Context class interact with Python and Clingo.
+    We can customize each value, range, and function to experiment on
+    different settings. Refer to the test case `test_small_agent_graph`
+    for more detail usages."""
+
+    number_of_leaves = 30
+    number_of_infra_agents = 20
+    number_of_product_agents = 60
+    infra_agent_profile = {"in_degree": [0, 10], "out_degree": [5, 100]}
+    product_agent_profile = {"in_degree": [5, 20], "out_degree": [0, 5]}
+
+    def nl(self) -> Symbol:
+        return Number(self.number_of_leaves)
+
+    def nia(self) -> Symbol:
+        return Number(self.number_of_infra_agents)
+
+    def npa(self) -> Symbol:
+        return Number(self.number_of_product_agents)
+
+    def infra_agent_out_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["out_degree"][0])
+
+    def infra_agent_out_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["out_degree"][1])
+
+    def infra_agent_in_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["in_degree"][0])
+
+    def infra_agent_in_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.infra_agent_profile["in_degree"][1])
+
+    def product_agent_out_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["out_degree"][0])
+
+    def product_agent_out_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["out_degree"][1])
+
+    def product_agent_in_degree_low(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["in_degree"][0])
+
+    def product_agent_in_degree_high(self, agent: Number) -> Symbol:
+        return Number(self.product_agent_profile["in_degree"][1])
+
+    def agent_connector(
```

Comment:
Please add a comment on what this function actually does, it's confusing from an outsider to review